File: 153-FindMinimumInRotatedSortedArray/153-FindMinimumInRotatedSortedArray.cpp
// Last updated: 16/10/2025, 04:28:51
class Solution {
public:
    int findMin(vector<int>& arr) {
        int n = arr.size();
        int s = 0;
        int e = n-1;
        int mid = (s + e)/2;
        
        while(s < e){
            if(arr[0] < arr[n-1]){
                return arr[0];
            }
            if(mid > 0 && arr[mid] < arr[mid-1]){
                return arr[mid];
            }
            else if(arr[mid] >= arr[0]){
                s = mid + 1;
            }
            else{
                e = mid;
            }
            
            mid = (s + e)/2;
        }
        //If last element is the sortest element
        return arr[e];
    }
};

File: 153-FindMinimumInRotatedSortedArray/153-FindMinimumInRotatedSortedArray.java
// Last updated: 16/10/2025, 04:28:45
class Solution {
    public int findMin(int[] nums) {
        int n = nums.length;
        int s = 0;
        int e = n-1;

        //Edge cases
        if(n == 1){   //size = 1
            return nums[0];
        }
        if(nums[0] < nums[n-1]){  //12345  not rotated
            return nums[0]; 
        }

        while(s < e){
            int mid = s + (e - s)/2;
            // Check if mid is the pivot
            if(mid - 1 >= 0 && nums[mid] < nums[mid - 1]){
             return nums[mid];
           }
           // Check if mid+1 is the pivot
           else if(mid + 1 <= n - 1 && nums[mid] > nums[mid + 1]){
            return nums[mid + 1];
           }
           //If nums[mid] >= nums[s], mid is on line s1 (4,5,...7)
           else if(nums[mid] >= nums[s]){
             s = mid + 1;
            //mid is on line s2 (0,1,2)
           }else{
             e = mid - 1;
          }
        }
        return -1; // Default return (shouldn't be reached in valid input)
    }
}